app/src/main/java/org/thoughtcrime/securesms/onboarding/LoadingViewModel.kt

```diff
@@ -0,0 +1,100 @@
+package org.thoughtcrime.securesms.onboarding
+
+import android.content.Context
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+import org.session.libsession.snode.SnodeModule
+import org.session.libsession.utilities.TextSecurePreferences
+import org.session.libsignal.database.LokiAPIDatabaseProtocol
+import org.session.libsignal.utilities.KeyHelper
+import org.session.libsignal.utilities.hexEncodedPublicKey
+import org.thoughtcrime.securesms.ApplicationContext
+import org.thoughtcrime.securesms.crypto.KeyPairUtilities
+import org.thoughtcrime.securesms.dependencies.ConfigFactory
+import javax.inject.Inject
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+data class State(val duration: Duration)
+
+private val ANIMATE_TO_DONE_TIME = 500.milliseconds
+private val IDLE_DONE_TIME = 1.seconds
+private val TIMEOUT_TIME = 15.seconds
+private val TOTAL_ANIMATION_TIME = TIMEOUT_TIME - IDLE_DONE_TIME
+
+@HiltViewModel
+class LoadingViewModel @Inject constructor(
+    private val configFactory: ConfigFactory,
+    private val prefs: TextSecurePreferences,
+) : ViewModel() {
+
+    private val state = MutableStateFlow(State(TOTAL_ANIMATION_TIME))
+    val stateFlow = state.asStateFlow()
+
+    private val event = Channel<Event>()
+    val eventFlow = event.receiveAsFlow()
+
+    private var restoreJob: Job? = null
+
+    internal val database: LokiAPIDatabaseProtocol
+        get() = SnodeModule.shared.storage
+
+    fun restore(context: Context, seed: ByteArray) {
+
+        // only have one sync job running at a time (prevent QR from trying to spawn a new job)
+        if (restoreJob?.isActive == true) return
+
+        restoreJob = viewModelScope.launch(Dispatchers.IO) {
+            // This is here to resolve a case where the app restarts before a user completes onboarding
+            // which can result in an invalid database state
+            database.clearAllLastMessageHashes()
+            database.clearReceivedMessageHashValues()
+
+            // RestoreActivity handles seed this way
```

Comment:
If there is shared logic, should this be centralised somewhere?

Reply:
@bemusementpark

Reply:
IDK, pre-existing code, not sure if there's subtle differences that do/don't matter.

Reply:
Yeah the idea is that we try to figure out those small things where we can, that's the only way we can clean up the app as we tackle various tickets.